src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs

```diff
@@ -933,54 +914,37 @@ private Task WriteToStreamAsync(ReadOnlyMemory<byte> source, CancellationToken c
         {
-            int searchOffset = 0;
+            int previouslyScannedBytes = 0;
             while (true)
             {
-                int startIndex = _readOffset + searchOffset;
-                int length = _readLength - startIndex;
-                int crPos = Array.IndexOf(_readBuffer, (byte)'\r', startIndex, length);
-                if (crPos < 0)
-                {
-                    // Couldn't find a \r.  Read more.
-                    searchOffset = length;
-                    await FillAsync(cancellationToken).ConfigureAwait(false);
-                }
-                else if (crPos + 1 >= _readLength)
+                int scanOffset = _readOffset + previouslyScannedBytes;
+                int endIndex = Array.IndexOf(_readBuffer, (byte)'\n', scanOffset, _readLength - scanOffset);
+                if (endIndex >= 0)
                 {
-                    // We found a \r, but we don't have enough data buffered to read the \n.
-                    searchOffset = length - 1;
-                    await FillAsync(cancellationToken).ConfigureAwait(false);
-                }
-                else if (_readBuffer[crPos + 1] == '\n')
-                {
-                    // We found a \r\n.  Return the data up to and including it.
-                    int lineLength = crPos - _readOffset + 2;
-                    var result = new ArraySegment<byte>(_readBuffer, _readOffset, lineLength);
-                    _readOffset += lineLength;
-                    return result;
-                }
-                else
-                {
-                    ThrowInvalidHttpResponse();
+                    int startIndex = _readOffset;
+                    _readOffset = endIndex + 1;
+
+                    if (endIndex > startIndex && _readBuffer[endIndex - 1] == '\r')
+                    {
+                        endIndex--;
+                    }
+
+                    return new ArraySegment<byte>(_readBuffer, startIndex, endIndex - startIndex);
                 }
-            }
-        }
 
-        private async Task ReadCrLfAsync(CancellationToken cancellationToken)
-        {
-            while (_readLength - _readOffset < 2)
-            {
-                // We have fewer than 2 chars buffered.  Get more.
+                // Couldn't find LF.  Read more.
+                // Note this may cause _readOffset to change.
+                previouslyScannedBytes = _readLength - _readOffset;
                 await FillAsync(cancellationToken).ConfigureAwait(false);
             }
+        }
 
-            // We expect \r\n.  If so, consume them.  If not, it's an error.
-            if (_readBuffer[_readOffset] != '\r' || _readBuffer[_readOffset + 1] != '\n')
+        private async Task ReadEmptyLineAsync(CancellationToken cancellationToken)
+        {
+            if (!LineIsEmpty(await ReadNextLineAsync(cancellationToken).ConfigureAwait(false)))
```

Comment:
While introducing ReadEmptyLineAsync that combines ReadNextLineAsync+LineIsEmpty improves composability, it also introduces cost.  If ReadNextLineAsync completes asynchronously, ReadEmptyLineAsync will itself incur the costs associated with yielding, i.e. the extra Task allocation (at least now with changes I made in 2.1 it's just that single allocation, though, rather than three or four).  Previously, the call to ReadNextLineAsync was essentially inlined into call sites that did the LineIsEmpty check, skipping the intermediate layer and thus the intermediate allocations.  Consider undoing this.

Reply:
Just to be clear -- are you saying we should get rid of ReadEmptyLineAsync and just do !LineIsEmpty(await ReadNextLineAsync)?

Reply:
Yup

Reply:
Ok, will do